### PR TITLE
pass exitcode to atexit hooks

### PR DESCRIPTION
**Pull request by simonbyrne**
Potential fix to #35981 and #47472.

cc @JeffBezanson @vchuravy

### CODE REVIEW

**Comment by DilumAluthge** [comment]
Would it also make sense to pass the signal number (if the Julia process is exiting because of a signal)?

**Comment by simonbyrne** [comment]
It would, but I'm not sure if we can access that?

**Comment by simonbyrne** [comment]
Ah, it is implicitly available if the exit code >=128
https://github.com/JuliaLang/julia/blob/cf9cac25e350522d26b341e1d9293b18030b9090/src/init.c#L236

**Review by vchuravy — APPROVED** [review]

File: base/initdefs.jl [hunk 1]
```diff
@@ -365,11 +373,15 @@ LIFO order, "last called" is equivalent to "first registered".)
 """
 atexit(f::Function) = (pushfirst!(atexit_hooks, f); nothing)
 
-function _atexit()
+function _atexit(exitcode::Cint)
     while !isempty(atexit_hooks)
         f = popfirst!(atexit_hooks)
         try
-            f()
+            if hasmethod(f, (Cint,))
```
[the remark this answers is not in the document]
Comment (simonbyrne) [reply base/initdefs.jl:380]: This was to make compatibility easier, since the you can define a hook with a default exit code

**Pull request merged**